### PR TITLE
[MultiThreading] New components and Task scheduler classes refactoring 

### PR DESCRIPTION
**Pull request by fspadoni**
Recover a commit in is_sofa_multithreading branch (3da9e50752d6aead487f0b5f8897bfc4249f9404) to allow OdeSolver concurrent execution:

A new data<bool> is added (d_threadsafevisitor) to disable allocation and writing in the mechanical object component used by the OdeSolver 
The same change is applied to all the OdeSolver derived class and it doesn't affect the default behavior and the scenes.
EulerSolver
EulerImplicitSolver
StaticSolver
CentralDifferenceSolver
VariationalSymplecticSolver
NewmarkImplicitSolver

Task Scheduler refactoring;
- improve task stealing between worker threads.
- add worker thread name to make the debugging and profiling easier.
- group and move the initialization tasks in a dedicated file.

New component and scene to show off the performance improvement of multithreading execution.
- MeanComputation

______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by epernod** [comment]
[ci-build][with-scene-tests]

File: SofaKernel/modules/SofaExplicitOdeSolver/EulerSolver.cpp [hunk 1]
```diff
@@ -59,6 +59,7 @@ SOFA_DECL_CLASS(Euler);
 
 EulerSolver::EulerSolver()
     : symplectic( initData( &symplectic, true, "symplectic", "If true, the velocities are updated before the positions and the method is symplectic (more robust). If false, the positions are updated before the velocities (standard Euler, less robust).") )
+    , d_threadsafevisitor(initData(&d_threadsafevisitor, false, "threadsafevisitor", "If true, do not use realloc and free visitors in fwdInteractionForceField."))
```
Comment (epernod) [inline SofaKernel/modules/SofaExplicitOdeSolver/EulerSolver.cpp]: just a quick remark, to follow Sofa coding rules. Could you replace the `d_threadsafevisitor ` by `d_threadsafeVisitor`  or `d_threadSafeVisitor` not sure which one is the most logical.

File: applications/plugins/MultiThreading/src/InitTasks.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            std::mutex*	 IdFactorygetIDMutex;
+            std::atomic<int>* _atomicCounter;
+        };
+
+
+        //fix and prefer using the global runThreadSpecificTask
+        SOFA_MULTITHREADING_PLUGIN_API void initThreadLocalData();
+        
+        
+#ifdef _WIN32
```
Comment (epernod) [inline applications/plugins/MultiThreading/src/InitTasks.h]: init and delete of OGL context is not needed on other platform?

File: applications/plugins/MultiThreading/src/MeanComputation.inl [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+                // get the lowest size
+                _resultSize = MechObjSizeMap.begin()->first;
+                d_result.beginEdit()->resize(_resultSize);
+
+                compute();
+            }
+
+            template <class DataTypes>
+            void MeanComputation<DataTypes>::bwdInit()
```
Comment (epernod) [inline applications/plugins/MultiThreading/src/MeanComputation.inl]: maybe clean or remove that method.

File: applications/plugins/MultiThreading/src/MeanComputation.inl [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                    }
+                }
+
+                for (int i = 0; i<_resultSize; ++i)
+                {
+                    result[i] *= invNbInputs;
+                }    
+                
+                d_result.endEdit();
+                d_result.setDirtyValue();
```
Comment (epernod) [inline applications/plugins/MultiThreading/src/MeanComputation.inl]: Normally it is not needed to setDirtyValue. 
Begin/endEdit automatically set dirty to true.

Could you test it without the line and tell me if it breaks something. For my understanding of Sofa Data.

File: SofaKernel/modules/SofaExplicitOdeSolver/EulerSolver.cpp (hunk 1 above)
Comment (epernod) [reply SofaKernel/modules/SofaExplicitOdeSolver/EulerSolver.cpp]: same in the other Solver classes

**Review by epernod — CHANGES_REQUESTED** [review]

File: SofaKernel/modules/SofaExplicitOdeSolver/EulerSolver.cpp (hunk 1 above)
Comment (fspadoni) [reply SofaKernel/modules/SofaExplicitOdeSolver/EulerSolver.cpp]: done commit 72d9ea24f2972f803113ba142599b391be429580.

File: applications/plugins/MultiThreading/src/MeanComputation.inl (hunk 3 above)
Comment (fspadoni) [reply applications/plugins/MultiThreading/src/MeanComputation.inl]: done

File: applications/plugins/MultiThreading/src/MeanComputation.inl (hunk 4 above)
Comment (fspadoni) [reply applications/plugins/MultiThreading/src/MeanComputation.inl]: Removed the d_result.setDirtyValue() and everything is fine.

**Review by epernod — APPROVED** [review]

**Comment by epernod** [comment]
[ci-build][with-scene-tests]

**Comment by epernod** [comment]
[ci-build][with-scene-tests]

**Comment by fspadoni** [comment]
[ci-build][with-scene-tests]

**Pull request merged**